pandas/core/arrays/timedeltas.py

```diff
@@ -137,25 +139,67 @@ def dtype(self):
     # Constructors
     _attributes = ["freq"]
 
-    @classmethod
-    def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
-        # `dtype` is passed by _shallow_copy in corner cases, should always
-        #  be timedelta64[ns] if present
-        assert dtype == _TD_DTYPE
-        assert isinstance(values, np.ndarray), type(values)
+    def __init__(self, values, dtype=_TD_DTYPE, freq=None, copy=False):
+        if isinstance(values, (ABCSeries, ABCIndexClass)):
+            values = values._values
+
+        if isinstance(values, type(self)):
+            if freq is None:
+                freq = values.freq
+            elif freq and values.freq:
+                freq = to_offset(freq)
+                freq, freq_infer = dtl.validate_inferred_freq(
+                    freq, values.freq,
+                    freq_infer=False
+                )
+            values = values._data
+
+        if not isinstance(values, np.ndarray):
+            msg = (
+                "Unexpected type '{}'. 'values' must be a TimedeltaArray "
+                "ndarray, or Series or Index containing one of those."
+            )
+            raise ValueError(msg.format(type(values).__name__))
 
         if values.dtype == 'i8':
-            values = values.view('m8[ns]')
-
-        assert values.dtype == 'm8[ns]'
-
-        result = object.__new__(cls)
-        result._data = values
-        result._freq = freq
-        return result
+            # for compat with datetime/timedelta/period shared methods,
+            #  we can sometimes get here with int64 values.  These represent
+            #  nanosecond UTC (or tz-naive) unix timestamps
+            values = values.view(_TD_DTYPE)
+
+        if values.dtype != _TD_DTYPE:
+            msg = (
+                "The dtype of 'values' is incorrect. Must be "
+                "'timedelta64[ns]'. Got '{}' instead."
+            )
+            raise ValueError(msg.format(values.dtype))
+
+        dtype_msg = "'dtype' must be 'timedelta64[ns]'. Got '{}' instead."
+        try:
+            if dtype != _TD_DTYPE:
+                raise ValueError(dtype_msg.format(dtype))
+        except TypeError:
+            raise ValueError(dtype_msg.format(dtype))
+
+        if freq == "infer":
+            msg = (
+                "Frequency inference not allowed in TimedeltaArray.__init__. "
+                "Use 'pd.array()' instead."
+            )
+            raise ValueError(msg)
+
+        if copy:
+            values = values.copy()
+        if freq:
+            freq = to_offset(freq)
```

Comment:
Not doing frequency validation here seems like an invitation to problems.  Users are going to assume `TimedeltaArray.__init__` is public and pass incorrect values here and we'll fail to raise.

Reply:
What's the consequence of passing an incorrect freq?

Reply:
> What's the consequence of passing an incorrect freq?

Methods that actually _use_ freq give nonsense answers, namely `shift`:

```
dti = pd.DatetimeIndex(['2018-12-27', '2018-12-28'])
dti._freq = pd.offsets.MonthBegin()  # <-- invalid frequency

>>> dti.shift(1)
DatetimeIndex(['2019-01-01'], dtype='datetime64[ns]', freq='MS')
```